core/runtime/src/main/java/io/quarkus/runtime/configuration/PropertiesUtil.java

```diff
@@ -121,23 +49,27 @@ public static boolean isPropertyQuarkusCompoundName(NameIterator propertyName) {
         return propertyName.getName().startsWith("\"quarkus.");
     }
 
-    /**
-     * Removes false positives of configuration properties marked as unknown. To populate the old @ConfigRoot, all
-     * properties are iterated and matched against known roots. With @ConfigMapping the process is different, so
-     * properties that are known to @ConfigMapping are not known to the @ConfigRoot, so they will be marked as being
-     * unknown. It is a bit easier to just double-check on the unknown properties and remove these false positives by
-     * matching them against the known properties of @ConfigMapping.
-     *
-     * @param unknownProperties the collected unknown properties from the old @ConfigRoot mapping
-     * @param filterPatterns the mapped patterns from the discovered @ConfigMapping
-     */
-    public static void filterUnknown(Set<String> unknownProperties, KeyMap<Boolean> filterPatterns) {
-        Set<String> toRemove = new HashSet<>();
-        for (String unknownProperty : unknownProperties) {
-            if (filterPatterns.hasRootValue(unknownProperty)) {
-                toRemove.add(unknownProperty);
+    public static Set<PropertyName> toPropertyNames(final Set<String> names) {
+        Map<PropertyName, String> propertyNames = new HashMap<>();
+        for (String name : names) {
+            PropertyName propertyName = new PropertyName(name);
+            if (propertyNames.containsKey(propertyName)) {
+                String existing = propertyNames.remove(propertyName);
+                if (existing.length() < name.length()) {
+                    propertyNames.put(new PropertyName(existing), existing);
+                } else if (existing.length() > name.length()) {
+                    propertyNames.put(propertyName, name);
+                } else {
+                    if (existing.indexOf('*') <= name.indexOf('*')) {
+                        propertyNames.put(new PropertyName(existing), existing);
+                    } else {
+                        propertyNames.put(propertyName, name);
+                    }
+                }
+            } else {
+                propertyNames.put(propertyName, name);
             }
         }
-        unknownProperties.removeAll(toRemove);
+        return propertyNames.keySet();
```

Comment:
I know that sounds paranoid (it is a bit I admit), but beware `keySet` if the lifecycle of the `String` used as values is to goes away: the key set within `HashMap` retain a strong reference to the original map (including the entries and values) (see https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/HashMap.java#L989 - the hidden `this` due to be an inner class)

Reply:
We can rewrite it, but I'm not very concerned. This is used only during build time, and the method calls are all local  when validating / generating the configuration, so everything should be GC by the end of the call.